evm/p2p/peer.py

```diff
@@ -350,36 +364,96 @@ def __init__(self, remote, privkey, reader, writer, aes_secret, mac_secret,
-                    # FIXME: The reference to the canonical chain head should not be stored in the
-                    # chaindb as that is shared by all peers -- instead, that should be kept on a
-                    # peer-specific db.
-                    self.chaindb.add_block_number_to_hash_lookup(header)
-                    self.chaindb.set_canonical_head(header.hash)
-                    head_number = header.block_number
-                self.logger.info("synced headers up to block {}".format(head_number))
+        if self.last_synced is None:
+            # It's the first time we hear from this peer, need to figure out which headers to
+            # get from it.  We can't simply fetch headers starting from our current head
+            # number because there may have been a chain reorg, so we fetch some headers prior
+            # to our head from the peer, and insert any missing ones in our DB, essentially
+            # making our canonical chain identical to the peer's up to
+            # chain_head.block_number.
+            chain_head = self.chaindb.get_canonical_head()
+            self.logger.info(
+                "First announcement by {}; head_number=={}, local chain_head=={}".format(
+                    self, head_number, chain_head.block_number))
+            oldest_ancestor_to_consider = max(
+                0, chain_head.block_number - self.max_headers_fetch + 1)
+            headers = yield from self._fetch_headers_starting_at(oldest_ancestor_to_consider)
+            if len(headers) == 0 or not self.chaindb.header_exists(headers[0].hash):
+                self.logger.warn("No common ancestors between us and remote, disconnecting")
+                self.disconnect(DisconnectReason.other)
+                return
+            for header in headers:
+                self.chaindb.persist_header_to_db(header)
+            self.last_synced = chain_head.block_number
+        else:
+            self.last_synced = self.last_synced - reorg_depth
+
+        while self.last_synced < head_number:
+            for header in (yield from self._fetch_headers_starting_at(self.last_synced + 1)):
+                self.chaindb.persist_header_to_db(header)
+                self.last_synced = header.block_number
+            self.logger.info("synced headers up to #{}".format(self.last_synced))
+
+    @asyncio.coroutine
+    def _fetch_headers_starting_at(self, start_block):
+        """Fetches up to self.max_headers_fetch starting at start_block.
+
+        Returns a tuple containing those headers in ascending order of block number.
+        """
+        request_id = gen_request_id()
+        self.les_proto.send_get_block_headers(
+            start_block, self.max_headers_fetch, request_id, reverse=False)
+        reply = yield from self.wait_for_reply(request_id)
+        self.logger.info("fetched headers from {} to {}".format(
+            reply['headers'][0].block_number, reply['headers'][-1].block_number))
```

Comment:
Is it possible for `reply['headers']` to be empty?

Reply:
In theory, yes, but that would indicate either a bug in our code (`_sync()` should never call this with a `start_block` that would cause an empty list of headers in the response) or in the remote if it announces a head higher than its actual one. I can add a check and raise an exception here so that at least we don't get a cryptic KeyError if that happens

Reply:
I'd feel safer with that check in place.

Reply:
I've added it already. Not super happy with the exception name, but maybe you can think of a better name?